src/geofabrics/bathymetry_estimation.py

```diff
@@ -0,0 +1,1576 @@
+        """ Return the number of samples to take along cross_sections. This should
+        be an odd number. Subtract 1 instead of adding to ensure within the
+        generated DEM. """
+
+        assert self.transect_radius is not None, "Transect radius must be set before this is called"
+        return int(self.transect_radius / self.resolution) * 2 - 1
+
+    def calculate_min_z_start_i(self, min_z_search_radius) -> int:
+        """ Return the starting index of samples along each transect to begin
+        looking for the minimum z.
+
+        Parameters
+        ----------
+
+        min_z_search_radius
+            The distance to search from the centre.
+        """
+
+        number_min_z_samples = int(min_z_search_radius / self.resolution) * 2 - 1
+
+        return int((self.number_of_samples - number_min_z_samples) / 2)
+
+    def calculate_min_z_stop_i(self, min_z_search_radius) -> int:
+        """ Return the stopping index of samples along each transect to begin
+        looking for the minimum z.
+
+        Parameters
+        ----------
+
+        min_z_search_radius
+            The distance to search from the centre.
+        """
+
+        return int(self.number_of_samples - self.calculate_min_z_start_i(min_z_search_radius))
+
+    @property
+    def centre_index(self) -> int:
+        """ Return the centre index for samples taken along a transect. """
+        return int(numpy.floor(self.number_of_samples / 2))
+
+    def _segment_slope(self, x_array, y_array, index):
+        """ Return the slope and length characteristics of a line segment.
+
+        Parameters
+        ----------
+
+        x_array
+            The x values of all polyline nodes.
+        y_array
+            The y values of all polyline nodes.
+        index
+            The segment index (the index of the starting node in the segment)
+        """
+        length = numpy.sqrt(
+            (x_array[index + 1] - x_array[index]) ** 2
+            + (y_array[index + 1] - y_array[index]) ** 2)
+        dx = (x_array[index + 1] - x_array[index]) \
+            / length
+        dy = (y_array[index + 1] - y_array[index]) \
+            / length
+        return dx, dy, length
+
+    def node_centred_reach_cross_section(self, sampled_channel: geopandas.GeoDataFrame):
+        """ Calculate cross_sections along a channel at the midpoint of each
+        segment.
+
+        Parameters
+        ----------
+
+        sampled_channel
+            The sampled channel defined as a single polyline. Any branches described
+            separately.
+        """
+
+        cross_sections_dict = {'geometry': [],
+                               'nx': [],
+                               'ny': [],
+                               'length': [],
+                               'mid_x': [],
+                               'mid_y': []}
+
+        assert len(sampled_channel) == 1, "Expect only one polyline " \
+            "geometry per channel. Instead got {len(channel_polyline)}"
```

Comment:
Applies to all asserts but SonarLint warns me:
 
"In production environments, the python -O optimization flag is often used, which bypasses assert statements."

Reply:
@rosepearson That sounds good to me, make sure to write a note or issue if you want to come back to it

Reply:
@LukeParky good point! I'll do this now. Thanks